packages/completer/src/handler.ts

```diff
@@ -500,6 +588,11 @@ export namespace CompletionHandler {
      * Any metadata that accompanies the completion reply.
      */
     metadata: ReadonlyJSONObject;
+
+    /**
+     * Hook for extensions to send ICompletionItems.
+     */
+    items?: ICompletionItems;
```

Comment:
While I like the idea in principle, this may require some extensive changes to existing downstream extensions. For instance, [jupyterlab-lsp](https://github.com/krassowski/jupyterlab-lsp/blob/master/packages/jupyterlab-lsp/src/adapters/jupyterlab/components/completion.ts#L32) exclusively uses a `LSPConnector` that inherits from `DataConnector` to communicate with JupyterLab.

Perhaps I could improve the documentation within `IReply` to make it less confusing?

Reply:
> For instance, jupyterlab-lsp exclusively uses a LSPConnector that inherits from DataConnector to communicate with JupyterLab.

It doesn't seem like that big a change for them. @krassowski do you have thoughts on this?

Is it true that if you provide `items` then the `matches` and `metadata` are unused? So a downstream user would have to just create dummy things to pass in there?

Reply:
I'm ok with either (I might slightly prefer @saulshanabrook suggestion). If the current form was to stay, should `matches` and `metadata` get deprecated?

Reply:
>  the fact that other packages also use the IDataConnector interface to fetch requests make me a little hesitant about introducing this new pattern.

I mean you could just use a DataConnector thing for the new param as well. But we only use the `fetch` function in it, so it just seems like an overkill abstraction IMHO but we do use it other places, so I would be fine reusing it here.

Reply:
I added `fetchItems` in `CompletionHandler.IOptions` as well as `ICompletionManager.ICompletableAttributes`, making `connector` optional. If `connector` is not set, I instantiate a `DummyConnector` that rejects the `fetch` call.

Reply:
Great! Thank you!